HeterogeneousCore/AlpakaInterface/src/alpaka/CachingAllocator.cc

```diff
@@ -0,0 +1,499 @@
+    auto& blocks = bin.blocks_;
+
+    std::vector<std::unique_ptr<BlockDescriptor>> temp;
+    temp.reserve(16);
+
+    bool found = false;
+    std::unique_ptr<BlockDescriptor> candidate;
+    while (not found and blocks.try_pop(candidate)) {
+      assert(candidate);
+      assert(candidate->buffer);
+      assert(candidate->buffer->data());
+      assert(candidate->queue);
+      assert(candidate->queue->m_spQueueImpl.get());
+      assert(candidate->event);
+      assert(candidate->event->m_spEventImpl.get());
+
+      if ((reuseSameQueueAllocations_ and (*block.queue == *candidate->queue)) or
+          alpaka::isComplete(*candidate->event)) {
+        // Reuse the block.
+        found = true;
+
+        // Transfer ownership of the old buffer.
+        block.buffer = std::move(candidate->buffer);
+        assert(block.buffer);
+        assert(block.buffer->data());
+
+        // Keep the queue used to request the allocation.
+        assert(block.queue);
+        assert(block.queue->m_spQueueImpl.get());
+
+        // If the new queue is on different device than the old event, create a
+        // new event, otherwise reuse the old event.
+        // Device memory allocators are per-device, so this can only happen for
+        // a host allocator caching mapped memory buffers.
+        if (block.device() != alpaka::getDev(*(candidate->event))) {
+          block.event = Event{block.device()};
+        } else {
+          if (debug_) {
+            // Only for debugging: ake a copy to print the information of the old event.
+            block.event = candidate->event;
+          } else {
+            block.event = std::move(candidate->event);
+          }
+        }
+        assert(block.event);
+        assert(block.event->m_spEventImpl.get());
+
+        // Update the accounting information.
+        totalFree_ -= block.bytes;
+        totalLive_ += block.bytes;
+        totalRequested_ += block.requested;
+
+        if (debug_) {
+          std::osyncstream out(std::cerr);
+          out << "\t" << deviceType_ << " " << alpaka::getName(device_) << " reused cached block at "
+              << block.buffer->data() << " (" << block.bytes << " bytes) for queue " << block.queue->m_spQueueImpl.get()
+              << ", event " << block.event->m_spEventImpl.get() << " (previously associated with queue "
+              << candidate->queue->m_spQueueImpl.get() << ", event " << candidate->event->m_spEventImpl.get() << ").\n";
+        }
+
+        // Free the block descriptor
+        candidate.reset();
+      } else {
+        // The candidate block is still busy, move it to the temporary buffer
+        temp.push_back(std::move(candidate));
+      }
+      assert(not candidate);
+    }
+
+    // Either we found a block to reuse, or the free list is empty; stop looking
+    // and put the blocks in the temporary store back in the free list.
+    for (auto& tmp_block : temp) {
+      blocks.push(std::move(tmp_block));
+    }
```

Comment:
> Do we need to worry about an exception being thrown between the `totalLive_` being increased on line 385, and the `allocate()` returning the `BlockDescriptor*` (that the called then encloses into a smart pointer)?

What can be done about it ?

Reply:
> > Do we need to worry about an exception being thrown between the `totalLive_` being increased on line 385, and the `allocate()` returning the `BlockDescriptor*` (that the called then encloses into a smart pointer)?
> 
> What can be done about it ?

It seems to me https://github.com/cms-sw/cmssw/pull/46658#discussion_r1893240614 shows a potential failure mode. `totalLive_` is increased here, but then the `alpaka::memset()` there throws an exception, leading to the memory held by `BlockDescriptor` released but `totalLive_` still accounting it, that should then lead to the `assert()` in the `CachingAllocator` destructor to fire.

One option would be to move the `totalLive_` (and maybe `totalRequested_` since those two seem to be modified always together) accounting to the end of `allocate()`, just before the `return`.

Reply:
> One option would be to move the `totalLive_` (and maybe `totalRequested_` since those two seem to be modified always together) accounting to the end of `allocate()`, just before the `return`.

Done.